File: src/lib/routing/bikeOverlayClass.test.ts
/**
 * Honesty: Scale-Farbe nur bei echten OSM-Tags; sac_scale ≠ S0–S3.
 * npx tsx src/lib/routing/bikeOverlayClass.test.ts
 */
import assert from "node:assert/strict";
import {
  classifyBikeWay,
  overlayFamilyForBike,
  overlayClassesForFamily,
  parseOsmMtbScale,
  bikeOverlaySurfaceKind,
} from "./bikeOverlayClass";

function testParseScale() {
  assert.equal(parseOsmMtbScale("2"), "S2");
  assert.equal(parseOsmMtbScale("S1"), "S1");
  assert.equal(parseOsmMtbScale("0+"), "S0");
  assert.equal(parseOsmMtbScale(undefined, "1"), "S1");
  assert.equal(parseOsmMtbScale("4"), "S3+");
  assert.equal(parseOsmMtbScale("S3"), "S3+");
  assert.equal(parseOsmMtbScale(""), null);
  assert.equal(parseOsmMtbScale(undefined, undefined), null);
}

function testMtbTagged() {
  const r = classifyBikeWay({
    highway: "path",
    "mtb:scale": "2",
  });
  assert.equal(r.bikeClass, "mtb");
  assert.equal(r.mtbScale, "S2");
}

function testImbaTagged() {
  const r = classifyBikeWay({
    highway: "track",
    "mtb:scale:imba": "0",
  });
  assert.equal(r.bikeClass, "mtb");
  assert.equal(r.mtbScale, "S0");
}

function testUnratedPathNotS2() {
  const r = classifyBikeWay({ highway: "path", surface: "ground" });
  assert.equal(r.bikeClass, "mtb_unrated");
  assert.equal(r.mtbScale, null);
}

function testSacScaleNotMapped() {
  const r = classifyBikeWay({
    highway: "path",
    sac_scale: "mountain_hiking",
    trail_visibility: "bad",
  });
  assert.equal(r.bikeClass, "mtb_unrated");
  assert.equal(r.mtbScale, null);
  assert.notEqual(r.mtbScale, "S2");
}

function testHiddenBicycleNo() {
  const r = classifyBikeWay({
    highway: "path",
    "mtb:scale": "1",
    bicycle: "no",
  });
  assert.equal(r.bikeClass, "hidden");
  assert.equal(r.mtbScale, null);
}

function testHiddenMtbNo() {
  const r = classifyBikeWay({ highway: "track", mtb: "no", surface: "ground" });
  assert.equal(r.bikeClass, "hidden");
}

function testGravelTrack() {
  const r = classifyBikeWay({
    highway: "track",
    tracktype: "grade3",
    surface: "compacted",
  });
  assert.equal(r.bikeClass, "gravel");
  assert.equal(r.mtbScale, null);
}

function testRoadCycleway() {
  const r = classifyBikeWay({
    highway: "cycleway",
    surface: "asphalt",
    bicycle: "designated",
  });
  assert.equal(r.bikeClass, "road");
}

function testUrbanLivingStreet() {
  const r = classifyBikeWay({ highway: "living_street" });
  assert.equal(r.bikeClass, "urban");
}

function testUrbanCycleLane() {
  const r = classifyBikeWay({
    highway: "residential",
    cycleway: "lane",
  });
  assert.equal(r.bikeClass, "urban");
}

function testFamilyDefaults() {
  assert.deepEqual(overlayClassesForFamily(overlayFamilyForBike("mtb_am")), [
    "mtb",
    "mtb_unrated",
  ]);
  assert.deepEqual(overlayClassesForFamily(overlayFamilyForBike("gravel")), [
    "gravel",
    "road",
  ]);
  assert.deepEqual(overlayClassesForFamily(overlayFamilyForBike("road")), [
    "road",
    "urban",
  ]);
  assert.deepEqual(overlayClassesForFamily(overlayFamilyForBike("urban")), [
    "urban",
    "road",
  ]);
  assert.equal(overlayFamilyForBike("emtb"), "mtb");
  assert.equal(overlayFamilyForBike("downhill"), "mtb");
  assert.equal(overlayFamilyForBike("dh"), "mtb");
}

function testSurfaceKindRideability() {
  assert.equal(bikeOverlaySurfaceKind("asphalt"), "paved");
  assert.equal(bikeOverlaySurfaceKind("concrete"), "paved");
  assert.equal(bikeOverlaySurfaceKind("compacted"), "gravel");
  assert.equal(bikeOverlaySurfaceKind("fine_gravel"), "gravel");
  assert.equal(bikeOverlaySurfaceKind("dirt"), "dirt");
  assert.equal(bikeOverlaySurfaceKind("ground"), "dirt");
  assert.equal(bikeOverlaySurfaceKind("unpaved"), "dirt");
  assert.equal(bikeOverlaySurfaceKind(""), "unknown");
  assert.equal(bikeOverlaySurfaceKind(undefined), "unknown");
  assert.equal(bikeOverlaySurfaceKind("woodchips"), "unknown");
}

testParseScale();
testMtbTagged();
testImbaTagged();
testUnratedPathNotS2();
testSacScaleNotMapped();
testHiddenBicycleNo();
testHiddenMtbNo();
testGravelTrack();
testRoadCycleway();
testUrbanLivingStreet();
testUrbanCycleLane();
testFamilyDefaults();
testSurfaceKindRideability();
console.log("bikeOverlayClass.test.ts: ok");

File: mobile/test/bike_overlay_class_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:aetherride_mobile/domain/bike.dart';
import 'package:aetherride_mobile/domain/routing/bike_overlay_class.dart';

void main() {
  test('mtb:scale path is mtb S2, never inferred from sac_scale', () {
    final tagged = classifyBikeWay({
      'highway': 'path',
      'mtb:scale': '2',
    });
    expect(tagged.bikeClass, BikeOverlayClass.mtb);
    expect(tagged.mtbScale, MtbScaleLabel.s2);
    expect(mtbScaleCss(MtbScaleLabel.s3), 'S3+');
    expect(parseOsmMtbScale('4'), MtbScaleLabel.s3);

    final sacOnly = classifyBikeWay({
      'highway': 'path',
      'sac_scale': 'mountain_hiking',
    });
    expect(sacOnly.bikeClass, BikeOverlayClass.mtbUnrated);
    expect(sacOnly.mtbScale, isNull);
  });

  test('untagged path/track is offen/unbewertet, not S2', () {
    final r = classifyBikeWay({'highway': 'path', 'surface': 'ground'});
    expect(r.bikeClass, BikeOverlayClass.mtbUnrated);
    expect(r.mtbScale, isNull);
  });

  test('hides bicycle=no and mtb=no', () {
    expect(
      classifyBikeWay({
        'highway': 'path',
        'mtb:scale': '1',
        'bicycle': 'no',
      }).bikeClass,
      BikeOverlayClass.hidden,
    );
    expect(
      classifyBikeWay({'highway': 'track', 'mtb': 'no'}).bikeClass,
      BikeOverlayClass.hidden,
    );
  });

  test('gravel / road / urban classes', () {
    expect(
      classifyBikeWay({
        'highway': 'track',
        'tracktype': 'grade2',
        'surface': 'fine_gravel',
      }).bikeClass,
      BikeOverlayClass.gravel,
    );
    expect(
      classifyBikeWay({
        'highway': 'cycleway',
        'surface': 'asphalt',
        'bicycle': 'designated',
      }).bikeClass,
      BikeOverlayClass.road,
    );
    expect(
      classifyBikeWay({'highway': 'living_street'}).bikeClass,
      BikeOverlayClass.urban,
    );
  });

  test('imba tag colors mtb, garage family defaults', () {
    final r = classifyBikeWay({
      'highway': 'track',
      'mtb:scale:imba': '0',
    });
    expect(r.bikeClass, BikeOverlayClass.mtb);
    expect(r.mtbScale, MtbScaleLabel.s0);
    expect(
      overlayClassesForFamily(overlayFamilyForBike(BikeCategory.mtbAm)),
      [BikeOverlayClass.mtb, BikeOverlayClass.mtbUnrated],
    );
    expect(
      overlayClassesForFamily(overlayFamilyForBike(BikeCategory.gravel)),
      [BikeOverlayClass.gravel],
    );
  });

  test('City default extraOn skips S-scale; MTB keeps trails', () {
    expect(
      overlayDefaultExtraOn(BikeOverlayFamily.urban),
      {BikeOverlayClass.urban, BikeOverlayClass.road},
    );
    expect(
      overlayDefaultExtraOn(BikeOverlayFamily.mtb),
      {BikeOverlayClass.mtb, BikeOverlayClass.mtbUnrated},
    );
    expect(overlayLegendShowsSScale(BikeOverlayFamily.urban), isFalse);
    expect(overlayLegendShowsSScale(BikeOverlayFamily.mtb), isTrue);
    expect(
      overlayLegendRows(family: BikeOverlayFamily.urban, expanded: false),
      isEmpty,
    );
    expect(
      overlayLegendRows(family: BikeOverlayFamily.urban, expanded: true)
          .map((r) => r.key),
      ['urban', 'road'],
    );
    expect(
      overlayLegendRows(family: BikeOverlayFamily.mtb, expanded: true)
          .map((r) => r.key),
      ['S0', 'S1', 'S2', 'S3+', 'unrated'],
    );
  });

  test('Discover overlay default is all classes; off hides, never 16%', () {
    expect(kAllPaintedOverlayClasses, contains(BikeOverlayClass.mtbUnrated));
    expect(kAllPaintedOverlayClasses, contains(BikeOverlayClass.road));
    expect(
      overlayClassesShown(
        overlayOn: true,
        extraOn: kAllPaintedOverlayClasses,
      ),
      kAllPaintedOverlayClasses,
    );
    expect(
      overlayClassesShown(
        overlayOn: false,
        extraOn: kAllPaintedOverlayClasses,
      ),
      isEmpty,
    );
    expect(
      overlayClassesShown(
        overlayOn: true,
        extraOn: {BikeOverlayClass.gravel},
      ),
      {BikeOverlayClass.gravel},
    );
  });

  test('signed cycle-route mesh keeps icn/ncn/rcn/mtb, drops lcn', () {
    expect(
      classifyBikeRoute({
        'route': 'bicycle',
        'network': 'icn',
        'ref': 'EV15',
      }).bikeClass,
      BikeOverlayClass.road,
    );
    expect(keepSignedCycleMesh({'route': 'bicycle', 'network': 'lcn'}), isFalse);
    expect(keepSignedCycleMesh({'route': 'bicycle', 'network': 'rcn'}), isTrue);
    expect(keepSignedCycleMesh({'route': 'bicycle', 'ref': 'EV6'}), isTrue);
    expect(keepSignedCycleMesh({'route': 'mtb'}), isFalse);
  });

  test('surface kind and line-color keep class fallback without field', () {
    expect(bikeOverlaySurfaceKind('asphalt'), BikeOverlaySurfaceKind.paved);
    expect(bikeOverlaySurfaceKind('compacted'), BikeOverlaySurfaceKind.gravel);
    expect(bikeOverlaySurfaceKind('dirt'), BikeOverlaySurfaceKind.dirt);
    expect(bikeOverlaySurfaceKind('unpaved'), BikeOverlaySurfaceKind.dirt);
    expect(bikeOverlaySurfaceKind(''), BikeOverlaySurfaceKind.unknown);
    final expr = bikeOverlaySurfaceLineColor(BikeOverlayColors.road);
    expect(expr.first, 'case');
    expect(expr.toString(), contains('has'));
    expect(expr.toString(), contains('surface'));
    expect(expr.toString(), contains('asphalt'));
    expect(expr.toString(), contains('any'));
    expect(expr[2], BikeOverlayColors.road);
    expect(expr.toString(), contains(BikeOverlayColors.dirt));
    expect(expr.toString(), contains(BikeOverlayColors.road));
  });
}
